Add "2ha.sh" script, managing 2-node Canonical K8s HA AA clusters (#692)
canonical/k8s-snap: docs/src/snap/howto/two-node-ha.md

Additional recovery steps are automated and invoked through Pacemaker.

### Prerequisites:

* Please ensure that both nodes are part of the Kubernetes cluster.
  See the [getting started] and [add/remove nodes] guides.
* The user associated with the HA service has SSH access to the peer node and
  passwordless sudo configured. For simplicity, the default "ubuntu" user can
  be used.
* We recommend using static IP configuration.

The [two-node-ha.sh script] automates most operations related to the two-node
HA scenario and is included in the snap.

The first step is to install the required packages:

```
/snap/k8s/current/k8s/hack/two-node-ha.sh install_packages
```

### Distributed Replicated Block Device (DRBD)

This example uses a loopback device as DRBD backing storage:

```
sudo dd if=/dev/zero of=/opt/drbd0-backstore bs=1M count=2000
```

Ensure that the loopback device is attached at boot time, before Pacemaker
starts.

```
cat <<EOF | sudo tee /etc/rc.local
#!/bin/sh
mknod /dev/lodrbd b 7 200
losetup /dev/lodrbd /opt/drbd0-backstore
EOF

sudo chmod +x /etc/rc.local
```

Add a service to automatically execute the ``/etc/rc.local`` script.

```

> 
> 

Reviewer: ```suggestion
```
Please remove these empty spaces

# systemd-rc-local-generator if /etc/rc.local is executable.
[Unit]
Description=/etc/rc.local Compatibility
Documentation=man:systemd-rc-local-generator(8)
ConditionFileIsExecutable=/etc/rc.local
After=network.target

[Service]
Type=forking
ExecStart=/etc/rc.local start
TimeoutSec=0
RemainAfterExit=yes
GuessMainPID=no

[Install]
WantedBy=multi-user.target
EOF

sudo systemctl enable rc-local.service
sudo systemctl start rc-local.service
```

Configure the DRBD block device that will hold the Dqlite data.
Please ensure that the correct node addresses are used.

```
# Disable the DRBD service, it will be managed through Pacemaker.
sudo systemctl disable drbd

HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

cat <<EOF | sudo tee /etc/drbd.d/r0.res
resource r0 {
  on haone {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HAONE_ADDR}:7788;
    meta-disk internal;
  }
  on hatwo {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HATWO_ADDR}:7788;
    meta-disk internal;
  }
}
EOF

sudo drbdadm create-md r0
sudo drbdadm status
```

Create a mount point for the DRBD block device. Non-default mount points

> need to be passed to the ``two-node-ha.sh`` script mentioned above, see the
> script for the full list of configurable parameters.

Reviewer: ```suggestion
need to be passed to the ``two-node-ha.sh`` script mentioned above. Please refer to the
script for the full list of configurable parameters.
```

```
DRBD_MOUNT_DIR=/mnt/drbd0
sudo mkdir -p $DRBD_MOUNT_DIR
```

> Run the following *once* to initialize the filesystem.

Reviewer: Does something bad happen if we run it twice that users should know about?

Author: Not really. It's just that unlike the other commands,  this doesn't have to be called twice.

service will not attempt to manage this device:

```
sudo cat <<EOF | sudo tee -a /etc/multipath.conf
blacklist {
    devnode "^drbd*"
}
EOF

sudo systemctl restart multipathd
```

### Corosync

Prepare the Corosync configuration. Again, make sure to use the correct
addresses.

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

sudo mv /etc/corosync/corosync.conf /etc/corosync/corosync.conf.orig

cat <<EOF | sudo tee /etc/corosync/corosync.conf
totem {
  version: 2
  cluster_name: ha
  secauth: off
  transport:udpu
  interface {
    ringnumber: 0
    bindnetaddr: ${netaddr}
    broadcast: yes
    mcastport: 5405
  }
}

nodelist {
  node {
    ring0_addr: ${HAONE_ADDR}
    name: haone
    nodeid: 1
  }
  node {
    ring0_addr: ${HATWO_ADDR}
    name: hatwo
    nodeid: 2
  }
}

quorum {
  provider: corosync_votequorum
  two_node: 1
  wait_for_all: 1
  last_man_standing: 1
  auto_tie_breaker: 0
}
EOF
```

Follow the above steps on both nodes before moving forward.

### Pacemaker

Let's define a Pacemaker resource for the DRBD block device, which
ensures that the block device will be mounted on the replica in case of a
primary node failure.

> [Pacemaker fencing] (stonith) configuration is environment specific and thus

Reviewer: ```suggestion
[Pacemaker fencing] (stonith) configuration is environment specific and thus
```
We should at least write out stonith once on the first mention.

> outside the scope of this guide. Using fencing is highly recommended if it is
> possible to reduce the risk of cluster split-brain situations.

Reviewer: ```suggestion
outside the scope of this guide. Using fencing is highly recommended, if it is
possible, to reduce the risk of cluster split-brain situations.
```
Commas here are optional but I believe this helps readability.

primitive drbd_res ocf:linbit:drbd params drbd_resource=r0 op monitor interval=29s role=Master op monitor interval=31s role=Slave
ms drbd_master_slave drbd_res meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true
primitive fs_res ocf:heartbeat:Filesystem params device=/dev/drbd0 directory=${DRBD_MOUNT_DIR} fstype=ext4
colocation fs_drbd_colo INFINITY: fs_res drbd_master_slave:Master
order fs_after_drbd mandatory: drbd_master_slave:promote fs_res:start
commit
show
quit
EOF
```

Before moving forward, let's ensure that the DRBD Pacemaker resource runs on
the primary (voter) Dqlite node.

In this setup, only the primary node holds the latest Dqlite data, which will
be transferred to the DRBD device once the clustered service starts.
This is automatically handled by the ``two-node-ha.sh start_service`` command.

```
sudo k8s status
sudo drbadadm status
sudo crm status
```

If the DRBD device is assigned to the secondary Dqlite node (spare), move it
to the primary like so:

```
sudo crm resource move fs_res <primary_node_name>

# remove the node constraint.
sudo crm resource clear fs_res
```

### Managing Kubernetes Snap Services

> For the two-node HA setup k8s snap services should no longer start

Reviewer: ```suggestion
For the two-node HA setup, k8s snap services should no longer start
```
This comma is not required either but helps the flow.

```
for f in `sudo snap services k8s  | awk 'NR>1 {print $1}'`; do
    echo "disabling snap.$f"
    sudo systemctl disable "snap.$f";
done
```

### Preparing the wrapper service

The next step is to define the wrapper service. Add the following to
``/etc/systemd/system/two-node-ha-k8s.service``.

```{note}
the sample uses the ``ubuntu`` user, feel free to use a different one as long as the prerequisites
are met.
```

```
[Unit]
Description=K8s service wrapper handling Dqlite recovery for two-node HA setups.
After=network.target pacemaker.service

[Service]
User=ubuntu
Group=ubuntu
Type=oneshot
ExecStart=/bin/bash /snap/k8s/current/k8s/hack/two-node-ha.sh start_service
ExecStop=/bin/bash sudo snap stop k8s
RemainAfterExit=true

[Install]
WantedBy=multi-user.target
```

```{note}
The ``two-node-ha.sh start_service`` command used by the service wrapper
automatically detects the expected Dqlite role based on the DRBD state.
It then takes the necessary steps to bootstrap the Dqlite state directories,
synchronize with the peer node (if available) and recover the database.
```

When a DRBD failover occurs, the ``two-node-ha-k8s`` service needs to be

> restarted. To accomplish this,, we are going to define a separate service that

Reviewer: ```suggestion
restarted. To accomplish this, we are going to define a separate service that
```
Okay now this second comma we don't need :)

``two-node-ha-k8s-failover`` service when a DRBD failover occurs.

```
sudo crm configure <<EOF
primitive ha_k8s_failover_service systemd:two-node-ha-k8s-failover op start interval=0 timeout=120 op stop interval=0 timeout=30
order failover_after_fs mandatory: fs_res:start ha_k8s_failover_service:start
colocation fs_failover_colo INFINITY: fs_res ha_k8s_failover_service
commit
show
quit
EOF
```

Once the setup is complete on both nodes, start the two-node HA k8s service on
each node:

```
sudo systemctl start two-node-ha-k8s
```

## Troubleshooting

Here are some potential problems that may affect two-node HA clusters and how
to address them.

### Dqlite recovery failing because of unexpected data segments

Dqlite recovery may fail if there are data segments past the latest snapshot.

```
Error: failed to recover k8s-dqlite, error: k8s-dqlite recovery failed, error:
recover failed with error code 1, error details: raft_recover(): io:
closed segment 0000000000002369-0000000000002655 is past last snapshot
snapshot-2-2048-642428, pre-recovery backup:
/var/snap/k8s/common/recovery-k8s-dqlite-2024-09-05T082644Z-pre-recovery.tar.gz
```

> Remove the offending segments and restart the ``two-node-ha-k8s`` service.

Reviewer: As a user I would wonder if it is save to remove this segment. Should we add an explanation? This is the data that has not been replicated to more than half of the voters and can therefore be safely removed?

Author: In our scenario, Dqlite doesn't replicate data at all, which is why we need DRBD. Removing data segments could lead to data loss, but there's no other documented way of recovering the cluster.

To be honest, the dqlite documentation is rather scarce and doesn't even mention this. I had to rely on the LXD documentation as well as the comments on GitHub issues:

* https://documentation.ubuntu.com/lxd/en/latest/howto/dqlite_troubleshoot/#interventions
* https://github.com/canonical/dqlite/issues/309#issuecomment-854736238

Reviewer: I see, can we add a warning similar to the one in LXD's docs please?

### DRBD split brain

The DRBD cluster may enter a [split brain] state and stop synchronizing. The
chances increase if fencing (stonith) is not enabled.

```
ubuntu@hatwo:~$ sudo drbdadm status
r0 role:Primary
  disk:UpToDate

ubuntu@hatwo:~$ cat /proc/drbd
version: 8.4.11 (api:1/proto:86-101)
srcversion: C7B8F7076B8D6DB066D84D9
 0: cs:StandAlone ro:Secondary/Unknown ds:UpToDate/DUnknown   r-----
    ns:0 nr:0 dw:0 dr:0 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:1802140

ubuntu@hatwo:~$ dmesg | grep "Split"
[  +0.000082] block drbd0: Split-Brain detected but unresolved, dropping connection!

> 

Reviewer: ```suggestion
```
Remove the extra line please.
